llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -478,51 +478,87 @@ Expected<uint64_t> materializeAbbrevFromTagImpl(MCCASReader &Reader,
   return Size + *MaybePaddingSize;
 }
 
+static Error materializeDebugInfoUnopt(MCCASReader &Reader,
+                                       ArrayRef<cas::ObjectRef> Refs,
+                                       SmallVectorImpl<char> &SectionContents) {
+
+  for (auto Ref : Refs) {
```

Comment:
(this comment also applies  to the first commit?)

I suspect it was easier to write this as a loop, but it seems we are trying to do some checking against ill-formed inputs. It would be a lot tighter, and easier to understand what we are checking, if we spelled out all the conditions explicitly:

* Check that the array has <=2 nodes, if not return an error.
* Check that the first node is a `DebugInfoUnoptRef`, if not return an error
 etc etc

The loop implementation will accept ill-formed inputs like a node with _only_ a PaddingRef (I understand you prevent this at the call site, but readers have to keep going back and forth between the two to realize that)